iceoryx_hoofs/include/iceoryx_hoofs/internal/posix_wrapper/mutex.hpp

```diff
@@ -17,81 +17,195 @@
 #ifndef IOX_HOOFS_POSIX_WRAPPER_MUTEX_HPP
 #define IOX_HOOFS_POSIX_WRAPPER_MUTEX_HPP
 
+#include "iceoryx_hoofs/cxx/expected.hpp"
 #include "iceoryx_hoofs/cxx/optional.hpp"
+#include "iceoryx_hoofs/design_pattern/builder.hpp"
 #include "iceoryx_hoofs/platform/pthread.hpp"
 
-#if defined(__QNX__) || defined(__APPLE__)
-#define PTHREAD_MUTEX_RECURSIVE_NP PTHREAD_MUTEX_RECURSIVE
-#define PTHREAD_MUTEX_FAST_NP PTHREAD_MUTEX_NORMAL
-#endif
-
 namespace iox
 {
 namespace posix
 {
+enum class MutexError
+{
+    INSUFFICIENT_MEMORY,
+    INSUFFICIENT_RESOURCES,
+    PERMISSION_DENIED,
+    INTER_PROCESS_MUTEX_UNSUPPORTED_BY_PLATFORM,
+    PRIORITIES_UNSUPPORTED_BY_PLATFORM,
+    USED_PRIORITY_UNSUPPORTED_BY_PLATFORM,
+    PRIORITY_MISMATCH,
+    MAXIMUM_NUMBER_OF_RECURSIVE_LOCKS_EXCEEDED,
+    DEADLOCK_CONDITION,
+    NOT_OWNED_BY_THREAD,
+    INVALID_PRIORITY_CEILING_VALUE,
+    UNDEFINED
+};
+
+enum class MutexTryLock
+{
+    LOCK_SUCCEEDED,
+    FAILED_TO_ACQUIRE_LOCK
+};
+
 /// @brief Wrapper for a interprocess pthread based mutex which does not use
 ///         exceptions!
 /// @code
 ///     #include "iceoryx_hoofs/internal/posix_wrapper/mutex.hpp"
 ///
 ///     int main() {
-///         posix::mutex myMutex(false);
+///         cxx::optional<iox::posix::Mutex> myMutex;
+///         iox::posix::MutexBuilder().isInterProcessCapable(true)
+///                                   .mutexType(MutexType::RECURSIVE)
+///                                   .priorityInheritance(MutexPriorityInheritance::NONE)
+///                                   .threadTerminationBehavior(MutexThreadTerminationBehavior::RELEASE_WHEN_LOCKED)
+///                                   .create(myMutex)
+///                                   .expect("Failed to create mutex!");
 ///
-///         myMutex->lock();
+///         myMutex->lock().expect("Mutex lock failed. Maybe the system is corrupted.");
 ///         // ... do stuff
-///         myMutex->unlock();
+///         myMutex->unlock().expect("Mutex unlock failed. Maybe the system is corrupted.");
 ///
 ///         {
-///             std::lock_guard<posix::mutex> lock(myMutex);
+///             std::lock_guard<posix::mutex> lock(*myMutex);
 ///             // ...
 ///         }
 ///
 ///     }
 /// @endcode
-/// @attention Errors in c'tor or d'tor can lead to a program termination!
-///
-class mutex
+class Mutex
 {
   public:
     /// @attention the construction of the mutex can fail. This can lead to a program termination!
-    explicit mutex(const bool f_isRecursive) noexcept;
+    /// @todo iox-#1036 remove this, introduced to keep current API temporarily
```

Comment:
This will be removed in the next PR. I kept it there to not be forced to integrate the new API into iceoryx which would blow up the PR.